influx-reports/influx.py

```diff
@@ -0,0 +1,241 @@
+    # Get points is generator, we should just get first
+    # point time (string type).
+    fpt_str = next(timestamp_query.get_points())['time']
+    try:
+        fpt = dateutil.parser.parse(
+            fpt_str) + datetime.timedelta(seconds=delta_seconds)
+    except Exception as alle:
+        raise InfluxTimestampParseException(
+            "Error parsing a timestamp from influx: {}".format(alle))
+    return fpt
+
+
+def calculate_measurement_uptime(client, measurement, job_reference,
+                                 measurement_period_seconds,
+                                 resolution=60):
+    """ For a certain build (job_reference), estimate the amount
+    of seconds a component (column of table) was up.
+    For that, we slice the job by resolution.
+    If I have no data in the slice, I assume downtime. If I have
+    at least one data per slice, I assume up, even if degraded.
+    The count of these positive events multiplied by resolution
+    is an average idea of the uptime.
+    """
+    per_field_up_percent=defaultdict(float)
+    measurement_uptime_percent = 0.0
+
+    query = (
+        "select max(/.*_status/) from {measurement} "
+        "where time < now() and job_reference='{job_reference}' "
+        "group by time({resolution}s) fill(-1)"
+    ).format(measurement=measurement,
+             job_reference=job_reference,
+             resolution=resolution)
+
+    max_amount_of_time_slices = math.ceil(float(measurement_period_seconds)/resolution)
+
+    all_data = client.query(query)
+    for time_slice in all_data.get_points():
+        measurement_was_up = False
+        for element in time_slice:
+            if element != 'time' and time_slice[element] == 1:
+                key = element.replace('max_', '').replace('_status','')
+                # Mark the field as up
+                per_field_up_percent[key] += (1 / max_amount_of_time_slices)
+                # Mark the measurement as up
+                measurement_was_up = True
+        if measurement_was_up:
+            measurement_uptime_percent += (1 / max_amount_of_time_slices)
+
+    return dict(
+        measurement_uptime_percent=measurement_uptime_percent,
+        per_field_uptime_percent=dict(per_field_up_percent)
+    )
+
+
+def build_report_dict(client, measurements, job_reference,
+                      measurement_period_seconds):
+    report = defaultdict(dict)
+    for measurement in measurements:
+        measurement_data = calculate_measurement_uptime(
+            client, measurement, job_reference, measurement_period_seconds)
+        report['runtime_seconds'] = measurement_period_seconds
+        report['measurement_uptime_percent'][measurement] = \
+            round(measurement_data['measurement_uptime_percent'] * 100, 4)
+        report['per_field_uptime_percent'][measurement] = \
+            dict((measure, round(measure_uptime * 100,4)) for measure, measure_uptime in measurement_data['per_field_uptime_percent'].items())
```

Comment:
At the very most round to two decimal places, even that is probably of no use but anything more is sub-second accuracy.